XMonad/Layout/FixedAspectRatio.hs

```diff
@@ -0,0 +1,123 @@
+--
+-- >      -- Set the aspect ratio of the focused window to 16:9
+-- >   ,((modm, xK_a), withFocused $ sendMessage . Fix (16 / 9))
+-- >
+-- >      -- Clear the aspect ratio from the focused window
+-- >   ,((modm .|. shiftMask, xK_a), withFocused $ sendMessage . Reset)
+--
+-- There's one caveat: to keep the usage of the modifier simple, it
+-- doesn't remove a window from its cache automatically. Which means
+-- that if you close a program window that has some fixed aspect ratios
+-- and relaunch it, sometimes it'll still have the fixed aspect ratio.
+-- You can try to avoid this by changing they keybinding used to kill
+-- the window:
+--
+-- >  , ((modMask .|. shiftMask, xK_c), withFocused (sendMessage . Reset) >> kill)
+--
+-- See "XMonad.Doc.Extending#Editing_key_bindings" for more info
+-- on customizing the keybindings.
+
+-- | Similar to 'layoutHintsWithReplacement', but relies on the user to
+-- provide the ratio for each window. @aspectRatio (rx, ry) layout@ will
+-- adapt the sizes of a layout's windows according to the provided aspect
+-- ratio, and position them inside their originally assigned area
+-- according to the @rx@ and @ry@ parameters.
+-- (0, 0) places the window at the top left, (1, 0) at the top right,
+-- (0.5, 0.5) at the center, etc.
+fixedAspectRatio
+  :: (Double, Double) -> l a -> ModifiedLayout FixedAspectRatio l a
+fixedAspectRatio = ModifiedLayout . FixedAspectRatio mempty
+
+data FixedAspectRatio a = FixedAspectRatio (M.Map Window Rational)
+                                           (Double, Double)
+  deriving (Read, Show)
+
+instance LayoutModifier FixedAspectRatio Window where
+  -- | Note: this resembles redoLayout from "XMonad.Layout.LayoutHints".
+  -- The only difference is relying on user defined aspect ratios, and
+  -- using the 'adj' function defined below instead of 'mkAdjust'
+  pureModifier (FixedAspectRatio ratios placement) _ (Just s) xs = (xs', Nothing)
+   where
+    xs' = map (\x@(_, r) -> second (placeRectangle placement r) $ applyHint x) xs
+    applyHint (win, r@(Rectangle x y w h)) =
+      let ar       = M.lookup win ratios
+          (w', h') = maybe (w, h) (adj (w, h)) ar
+      in  (win, if isInStack s win then Rectangle x y w' h' else r)
+
+  pureModifier _ _ _ xs = (xs, Nothing)
+
+  handleMess (FixedAspectRatio ratios placement)  mess = case fromMessage mess of
+    Just (Fix r w) -> return . Just $ FixedAspectRatio (M.insert w r ratios) placement
+    Just (Reset w) -> return . Just $ FixedAspectRatio (M.delete w ratios) placement
+    _              -> return Nothing
+
+
+-- | Calculates the new width and height so they respect the
+-- aspect ratio.
+adj :: (Dimension, Dimension) -> Rational -> (Dimension, Dimension)
+adj (w, h) ar | ar' > ar  = (ceiling $ fi h * ar, h)
+              | otherwise = (w, ceiling $ fi w / ar)
+  where ar' = fi w % fi h
+
+--- Message handling
+data ManageAspectRatio =
+    Fix Rational Window -- ^ Set the aspect ratio for the window
```

Comment:
If this module is meant to bude used unqualified (as most modules probably are), I'd consider `FixAspect` and `ResetAspect` to be clearer and less likely to conflict with other modules.

Reply:
Unrelated idea: maybe a managehook helper `doFixAspect` might be a useful addition. Should probably use `sendMessageWithNoRefresh` to avoid refreshing before the window gets managed.

Reply:
I like the idea with the manageHook :+1: